snuba/subscriptions/subscription.py

```diff
@@ -22,12 +22,13 @@ class SubscriptionCreator:
     the resulting query is valid.
     """
 
-    def __init__(self, dataset: Dataset):
+    def __init__(self, dataset: Dataset, entity_key: EntityKey):
         self.dataset = dataset
-        self.entity_key = ENTITY_NAME_LOOKUP[dataset.get_default_entity()]
+        self.entity_key = entity_key
```

Comment:
This is going to be a problem and a much larger refactoring.
The entire subscription infra works only with one default entity per dataset. Now you need to create subscriptions on multiple entities on the same dataset. 
The problems you are encountering is that: there is no in code relationship between entity and dataset. We still did not build that. So you may create a subscription for entity `metrics_counter` and dataset `discover` and this will be fine. Except that everything will break when you try to execute the query. If you are lucky that would be noticed by the `_test_request` method here, but you should not rely on that. @evanh  do we have a  task anywhere to store this relationship in code eventually ? I think it will have to be done before this PR.

Reply:
I think at a minimum you could just validate that the entity passed corresponds to one that is in `dataset.get_all_entities()`

Reply:
fwiw I did create that relationship between dataset and entity on the sentry side so technically its not possible atm to `discover` dataset and `metrics_counters` entity.
Here i a link to that: https://github.com/getsentry/sentry/pull/30400/files#diff-b3812aba29e6dd5fb20c8e98e849b988faf38564c8f3dcd5526bab213c8f7853R238-R274

However, I agree with @lynnagara I should probably also validate this using `dataset.get_all_entities()` so I added that change